enhancements/proxy-passdown.md

```diff
@@ -0,0 +1,162 @@
+---
+title: proxy-friendly-operators
+authors:
+  - "@asmacdo"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-07-02
+last-updated: 2021-07-02
+status: implementable
+---
+
+# Operators Support Network Proxy
+
+## Summary
+
+Document, simplify, and encourage Operator authors to pass proxy
+environment variables to operands.
+
+## Motivation
+
+Working with a proxy is a feature that needs to be added to each
+operator separately, enabling the end-user story:
+"As an operator-admininstrator, I
+can deploy operators that respect a network proxy."
+
+For Cluster administrators to set up a network proxy:
+   1. Pods requiring network must respect proxy environment variables
+      `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY`
+   2. Operators must be provided environment variables
+   3. Operators must pass the environment variables to operands
+
+All of this is already possible but operator authors must implement this
+functionality themselves. This should be a "best practice"
+and documentation for this in our repository.
+
+## Proposal
+
+### Goals
+
+   1. Establish a best practice for operator authors:  Operators receive proxy
+      environment variables from the operator deployment, and pass proxy
+      environment variables to operands that require network.
+   2. Operator SDK features and documents to facilitate implementation of
+      proxy-friendly operators.
+
+### Non-Goals
+
+   1. It would be nice to encourage proxy friendly operators by
+      including that functionality as a badge or part of the maturity
+      model on operatorhub, but that is out of scope for this proposal.
+   2. Automatic scaffolding of proxy passdown cannot be done because the
+      logic must occur in operator-specific code.
```

Comment:
Is it because the injected codes depend on the type of k8s resources being used by Operand instance? **_e.g._** it could be a `Deployment` or a `StatefulSet`. 

If that's the case, can we perhaps default scaffold out sample codes with `Deployment` and we provide other helpers for `StatefulSet` and document how to change to `StatefulSet` if `Deployment` is not what they want? 

(I agree the scaffolding piece could come later or outside of this EP, just would like to know better if that's attainable later even if it's not for now.)

Reply:
Thats part of it. Here's what we currently scaffold for the go reconcile:

```go
func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	_ = log.FromContext(ctx)

	// your logic here

	return ctrl.Result{}, nil
}
```
AIUI our current strategy is to scaffold only the skeleton, and not to include example dummy code, so I think scaffolding a suggested best practice for one of many possible directions would be awkward, but I'm not opposed to it.

Reply:
What we can do instead is include this in the tutorial, which implements this function. https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/#reconcile-loop

`operatorlib.AddProxyToDeployment(memcached)  // pass along proxy env vars`

Reply:
> AIUI our current strategy is to scaffold only the skeleton, and not to include example dummy code, so I think scaffolding a suggested best practice for one of many possible directions would be awkward, but I'm not opposed to it.

Agree this is worthy of a broader scope discussion and not in the scope of this EP.

Reply:
> What we can do instead is include this in the tutorial, which implements this function. https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/#reconcile-loop
> 
> `operatorlib.AddProxyToDeployment(memcached) // pass along proxy env vars`

Adding this to the tutorial to promote this sounds good to me, thanks!